pmm-app/src/pmm-dbaas/components/XtraDB/XtraDB.hooks.ts

```diff
@@ -0,0 +1,44 @@
+import { useState, useEffect } from 'react';
+import { processPromiseResults, filterFulfilled, FulfilledPromiseResult } from 'shared/components/helpers/promises';
+import { DATABASE_LABELS } from 'shared/core';
+import { Kubernetes } from '../Kubernetes/Kubernetes.types';
+import { XtraDBCluster, GetXtraDBClustersAction } from './XtraDB.types';
+import { XtraDBService, toModel } from './XtraDB.service';
+
+export const useXtraDBClusters = (kubernetes: Kubernetes[]): [
+  XtraDBCluster[],
+  GetXtraDBClustersAction,
+  boolean
+] => {
+  const [xtraDBClusters, setXtraDBClusters] = useState<XtraDBCluster[]>([]);
+  const [loading, setLoading] = useState(false);
+
+  const getXtraDBClusters = async () => {
+    setLoading(true);
+
+    try {
+      const clusters: XtraDBCluster[] = [];
+      const requests = kubernetes.map(XtraDBService.getXtraDBClusters);
+      const results = await processPromiseResults(requests);
+
+      results
+        .filter(filterFulfilled)
+        .map((r) => ((r as FulfilledPromiseResult).value.clusters)
+          .map(
```

Comment:
I'd say that using a push inside the map is not the best option, something like this looks simpler

` const clusters = results.filter(filterFulfilled).map(({value}) => toModel(value.clusters, DATABASE_LABELS.mysql)) `

Reply:
That wouldn't work, because "value.clusters" is a list of clusters and not a single cluster. We have to do one request per kubernetes cluster to get the DB clusters for that kubernetes cluster. That gives us a list of lists. A list of DB clusters per kubernetes cluster. I'm using the push there because in the end we want a single list and not a list of lists. Doing it this way prevents us from having to further iterate the list again to do that transformation. Tell me if you understand it @lunaticusgreen and if still think there's a better way pls.

Reply:
got it, thanks for the explanation 👍